File: zig/day04/board.zig
const std = @import("std");
const util = @import("../util.zig");
const Self = @This();

pub const Num = std.math.IntFittingRange(0, 99);
const Mask = std.meta.Int(.unsigned, SIZE * SIZE);
const Shift = std.meta.Int(.unsigned, SIZE);

const SIZE = 5;
const ROW_MASK: Mask = 0b00000_00000_00000_00000_11111;
const COL_MASK: Mask = 0b00001_00001_00001_00001_00001;

matrix: [SIZE * SIZE]Num,
marked: Mask,

pub fn parse(input: []const u8) !Self {
  var self = std.mem.zeroes(Self);
  var nums = std.mem.tokenize(u8, input, " \n\r");
  var i: usize = 0;
  while (nums.next()) |num| : (i += 1)
    self.matrix[i] = try std.fmt.parseUnsigned(Num, num, 10);
  return self;
}

pub fn mark(self: *Self, num: Num) void {
  for (self.matrix) |n, i| if (n == num) {
    self.marked |= @as(Mask, 1) << @intCast(Shift, i);
    return;
  };
}

pub fn sumUnmarked(self: Self) u32 {
  var sum: u32 = 0;
  for (self.matrix) |n, i| {
    if (self.marked & @as(Mask, 1) << @intCast(Shift, i) == 0)
      sum += n;
  }
  return sum;
}

pub fn bingo(self: Self) bool {
  var i: Shift = 0;
  while (i < SIZE) : (i += 1) {
    const row = ROW_MASK << i * SIZE;
    const col = COL_MASK << i;
    if (self.marked & row == row or self.marked & col == col)
      return true;
  }
  return false;
}

File: rust/src/day04/board.rs
pub const N: usize = 5;

const ROW_MASK: u32 = 0b_00000_00000_00000_00000_11111;
const COL_MASK: u32 = 0b_00001_00001_00001_00001_00001;

pub struct Board {
  matrix: [u8; N * N],
  marked: u32
}

impl Board {
  pub fn parse(input: &str) -> Option<Self> {
    let mut matrix = [0; N * N];
    for (num, cell) in input.split_ascii_whitespace().zip(&mut matrix) {
      *cell = num.parse().ok()?;
    }
    Some(Self { matrix, marked: 0 })
  }

  pub fn at(&self, col: usize, row: usize) -> u8 {
    self.matrix[N * row + col]
  }

  pub fn is_marked_at(&self, col: usize, row: usize) -> bool {
    self.marked & 1 << (N * row + col) != 0
  }

  pub fn mark(&mut self, num: u8) {
    if let Some(shift) = self.matrix.iter().position(|&n| n == num) {
      self.marked |= 1 << shift;
    }
  }

  pub fn sum_all_unmarked(&self) -> u32 {
    self.matrix.iter().enumerate()
      .filter(|&(shift, _)| self.marked & 1 << shift == 0)
      .fold(0, |acc, (_, &n)| acc + n as u32)
  }

  pub fn bingo(&self) -> bool {
    for i in 0..N {
      let row = ROW_MASK << i * N;
      let col = COL_MASK << i;
      if self.marked & row == row || self.marked & col == col {
        return true;
      }
    }
    false
  }
}
